docs/architecture/adr-017-single-binary-upgrade.md

```diff
@@ -0,0 +1,157 @@
+  - Example: Osmosis is currently on [version.app: 15](https://www.mintscan.io/osmosis/blocks/8804111) and release [v15.0.0](https://github.com/osmosis-labs/osmosis/releases/tag/v15.0.0).
+
+  - Have to maintain old versions of binaries. Need to update dependencies for security issues.
+
+- Genesis has an original version.app. Does celestia-app need to explicitly define this?
+
+- Bitcoin / Ethereum leverage if statement upgrade logic. Downside to if statement logic in Cosmos ecosystem is if statements may not be relevant for new chains
+
+- UX issue to improve experience for node operators to upgrade
+
+- Maintenance issue for maintaining historical releases of binaries
+
+- What granularity should conditional logic be?
+  - App struct level
+  - Module level
+  - if/else on a line by line basis
+
+- Why app version and not block height?
+  - App version is bumped via upgrade module
+  - Depending on app version is universal across multiple networks.
+  - Block height will differ across testnets / chain-ids so depends on a `.toml` config file
+
+- Are we planning on upstream this?
+  - [CometBFT](https://github.com/cometbft/cometbft) is working on this for consensus.
+  - We’re working on Cosmos SDK portion
+
+- `celestia-node` light nodes need to version erasure coding block or celestia-node should import this from celestia-app
+
+- Will celestia-node need to pass a block height to celestia-app? Header contains version.app (state machine version) and version.block (consensus version)
+
+- Is it important to support rolling back versions? Involves upgrading (again) through the faulty version
+
+- Since if conditional logic uses one Go version, all code will be compiled with this one Go version. If behavior changes across Go versions, we have to debug why.
+
+- Option: depend on multiple major versions of a dependency. What is the relationship between if/else conditional statements and Cosmos SDK upstream changes?
+
+- How to do this in CometBFT?
+  - Always use latest p2p version
+  - Version.block in header should be used to version changes to Tendermint / CometBFT
+
+- Block sync needs to be aware of changes across block heights
+
+### Implementation proposal
+
+There are multiple ways of doing it. The app version can be found in the SDK context.
+So, in the `PrepareProposal` method, we can get the app version [from the context](https://github.com/celestiaorg/celestia-app/blob/main/app/prepare_proposal.go#L28) and use it as the version indicator.
+
+#### Option 1: Passing the app version throughout all the required functions
+
+In this case we need to pass either sdk context or the app version itself to all the functions and objects that we think they might be affected by a version change in future.
+So for those functions that do not have the ctx already, we need to add an extra parameter to pass it.
```

Comment:
which functions do not have access to the app version? I believe just PrepareProposal, as the header is passed to ProcessProposal